nautobot_ssot/jobs/base.py

```diff
@@ -33,31 +36,124 @@
 
     def sync_data(self):
-        """Method to be implemented by data sync concrete Job implementations.
+        """Method to load data from adapters, calculate diffs and sync (if not dry-run).
+
+        It is composed by 4 methods:
+        - self.load_source_adapter: instantiates the source adapter (self.source_adapter) and loads its data
+        - self.load_target_adapter: instantiates the target adapter (self.target_adapter) and loads its data
+        - self.calculate_diff: generates the diff from source to target adapter and stores it in self.diff
+        - self.execute_sync: if not dry-run, uses the self.diff to synchronize from source to target
 
+        This is a generic implementation that you could overwrite completely in you custom logic.
         Available instance attributes include:
 
         - self.kwargs     (corresponds to the Job's `data` input, including 'dry_run' option)
         - self.commit     (should generally be True)
         - self.sync       (Sync instance tracking this job execution)
         - self.job_result (as per Job API)
         """
-        pass  # pylint: disable=unnecessary-pass
+        if not self.sync:
+            return
+
+        if self.kwargs["memory_profiling"]:
+            tracemalloc.start()
+
+        start_time = datetime.now()
+
+        self.log_info(message="Loading current data from source adapter...")
+        self.load_source_adapter()
+        load_source_adapter_time = datetime.now()
+        self.sync.load_time_source = load_source_adapter_time - start_time
+        self.sync.save()
+        self.log_info(message=f"Source Load Time from {self.source_adapter}: {self.sync.load_time_source}")
+
+        self.log_info(message="Loading current data from target adapter...")
+        self.load_target_adapter()
+        load_target_adapter_time = datetime.now()
+        self.sync.load_time_target = load_target_adapter_time - load_source_adapter_time
+        self.sync.save()
+        self.log_info(message=f"Target Load Time from {self.target_adapter}: {self.sync.load_time_target}")
+
+        self.log_info(message="Calculating diffs...")
+        self.calculate_diff()
+        calculate_diff_time = datetime.now()
+        self.sync.diff_time = calculate_diff_time - load_target_adapter_time
+        self.sync.save()
+        self.log_info(message=f"Diff Calculation Time: {self.sync.diff_time}")
+
+        if not self.kwargs["dry_run"]:
+            self.log_info(message=f"Syncing from {self.source_adapter} to {self.target_adapter}...")
+            self.execute_sync()
+            execute_sync_time = datetime.now()
+            self.sync.sync_time = execute_sync_time - calculate_diff_time
+            self.sync.save()
+            self.log_info(message="Sync complete")
+            self.log_info(message=f"Sync Time: {self.sync.sync_time}")
+
+        if self.kwargs["memory_profiling"]:
+            memory_size, memory_peak = tracemalloc.get_traced_memory()
```

Comment:
Any reason not to save these two values to `self.sync` as well? Are they less useful?

Reply:
fixed